src/librustc_typeck/check/never_compat.rs

```diff
@@ -0,0 +1,433 @@
+    ///
+    /// We look for a type variable that ended up unified with a diverging
+    /// inference variable. This represents a place where fallback to a never
+    /// type ended up affecting a live method call. We then return the
+    /// inference variable, along with the corresponding divering inference
+    /// variable that was unified with it
+    ///
+    /// Note that their be multiple such pairs of inference variables - howver,
+    /// we only display one to the user, to avoid overwhelming them with information.
+    fn find_best_vars(&self, fcx: &FnCtxt<'a, 'tcx>, vars: &[Ty<'tcx>]) -> VarData {
+        for var in vars {
+            for diverging_var in &self.unconstrained_diverging {
+                match (&var.kind, &diverging_var.kind) {
+                    (ty::Infer(ty::InferTy::TyVar(vid1)), ty::Infer(ty::InferTy::TyVar(vid2))) => {
+                        if fcx.infcx.type_variables.borrow_mut().sub_unified(*vid1, *vid2) {
+                            debug!(
+                                "Type variable {:?} is equal to diverging var {:?}",
+                                var, diverging_var
+                            );
+
+                            debug!(
+                                "Var origin: {:?}",
+                                fcx.infcx.type_variables.borrow().var_origin(*vid1)
+                            );
+                            return VarData { best_var: *vid1, best_diverging_var: *vid2 };
+                        }
+                    }
+                    _ => bug!("Unexpected types: var={:?} diverging_var={:?}", var, diverging_var),
+                }
+            }
+        }
+        bug!("No vars were equated to diverging vars: {:?}", vars)
+    }
+
+    /// The hook used by typecheck after fallback has been run.
+    /// We look for any "questionable" calls, generating diagnostic
+    /// message for them.
+    pub fn post_fallback(self, fcx: &FnCtxt<'a, 'tcx>) {
+        let tcx = fcx.tcx;
+        for (call_id, path) in &self.unresolved_paths {
+            debug!(
+                "post_fallback: resolved ty: {:?} at span {:?} : expr={:?} parent={:?} path={:?}",
+                path.span,
+                path.ty,
+                tcx.hir().get(*call_id),
+                tcx.hir().get(tcx.hir().get_parent_node(*call_id)),
+                path
+            );
+
+            let span = path.span;
+            if let Some(vars) = find_questionable_call(path, fcx) {
+                let VarData { best_var, best_diverging_var } = self.find_best_vars(fcx, &vars);
+
+                let var_origin = *fcx.infcx.type_variables.borrow().var_origin(best_var);
+                let diverging_var_span =
+                    fcx.infcx.type_variables.borrow().var_origin(best_diverging_var).span;
+
+                let mut err = tcx
+                    .sess
+                    .struct_span_err(span, "Fallback to `!` may introduce undefined behavior");
```

Comment:
```suggestion
                    .struct_span_warn(span, "fallback to `!` may introduce undefined behavior");
```

Reply:
I'm happy with keeping this an error for the purposes of a crater run, but this needs to become a warning (or a lint, depending on your opinions re. `--cap-lints`) before the PR can land.